### PR TITLE
Feat: Changes config to add support for bytes input/return types

### PR DESCRIPTION
**Pull request by Raiden1411**
## Description

Aims to add support for what was discussed on #77. 

@tmm let me know if something like this works or if you had something else in mind.

## Additional Information

- [x] I read the [contributing docs](/wagmi-dev/abitype/blob/main/.github/CONTRIBUTING.md) (if this is your first contribution)

Your ENS/address: 0x75EF32C862b8fe2915D49b38a2AECc8744622864

### CODE REVIEW

File: src/utils.ts [hunk 1]
```diff
@@ -75,10 +82,11 @@ type BitsTypeLookup = {
  */
 export type AbiParameterToPrimitiveType<
   TAbiParameter extends AbiParameter | { name: string; type: unknown },
+  TAbiParameterType extends 'inputs' | 'outputs' = 'inputs',
```
Comment (tmm) [inline src/utils.ts:85]: See comment below on `AbiParametersToPrimitiveTypes`

File: src/config.ts [hunk 2]
```diff
@@ -80,13 +82,22 @@ export interface ResolvedConfig {
   AddressType: IsUnknown<Config['AddressType']> extends true
     ? DefaultConfig['AddressType']
     : Config['AddressType']
+
+  /**
+   * TypeScript type to use for `bytes` values
+   * @default `0x${string}`
+   */
+  BytesInputType: IsUnknown<Config['BytesInputType']> extends true
```
Comment (tmm) [inline src/config.ts:90]: Another idea for the config type is keeping the name `BytesType`, but allowing it to be an object:

```ts
export interface DefaultConfig {
  BytesType: {
    Inputs: `0x${string}` | Uint8Array
    Outputs: `0x${string}`
  }
}
```

**Review by tmm — COMMENTED** [review]
Looking solid. Left a few comments. You'll also want to add a [changeset](https://github.com/wagmi-dev/abitype/blob/main/.github/CONTRIBUTING.md#versioning).

File: src/config.ts (hunk 2 above)
Comment (Raiden1411) [reply src/config.ts:90]: I like this approach better. I will go with this. Will need to remove the capital lettering of Inputs and Outputs in order to keep with the abi spec as well as using `AbiParameterType` as keys for the object. This approach also works best with the `TypedData` types. 

Will also update the read me to reflect this change.

File: src/utils.ts [hunk 3]
```diff
@@ -164,11 +177,15 @@ type _HasUnnamedAbiParameter<TAbiParameters extends readonly AbiParameter[]> =
  */
 export type AbiParametersToPrimitiveTypes<
   TAbiParameters extends readonly AbiParameter[],
+  TAbiParameterType extends 'inputs' | 'outputs' = 'inputs',
```
[the remark this answers is not in the document]
Comment (Raiden1411) [reply src/utils.ts:180]: I agree. I initially decided to make it a required generic but it felt too declarative/enforcing. Having the default be the union is a better approach than just the `inputs` literal. 

Will update the generic and add the new type.

File: src/config.ts [hunk 4]
```diff
@@ -80,13 +82,22 @@ export interface ResolvedConfig {
...
+    ? DefaultConfig['BytesInputType']
+    : Config['BytesInputType']
+
   /**
    * TypeScript type to use for `bytes` values
```
[the remark this answers is not in the document]
Comment (Raiden1411) [reply src/config.ts:95]: Check comments bellow.

File: src/config.ts [hunk 5]
```diff
@@ -80,13 +82,22 @@ export interface ResolvedConfig {
   AddressType: IsUnknown<Config['AddressType']> extends true
     ? DefaultConfig['AddressType']
     : Config['AddressType']
+
+  /**
+   * TypeScript type to use for `bytes` values
+   * @default `0x${string}`
```
[the remark this answers is not in the document]
Comment (Raiden1411) [reply src/config.ts:88]: Check comments bellow.

File: src/config.ts [hunk 6]
```diff
@@ -27,8 +27,10 @@ export interface DefaultConfig {
 
   /** TypeScript type to use for `address` values */
   AddressType: `0x${string}`
-  /** TypeScript type to use for `bytes` values */
-  BytesType: `0x${string}`
+  /** TypeScript type to use for `bytes` return values */
+  BytesReturnType: `0x${string}`
```
[the remark this answers is not in the document]
Comment (Raiden1411) [reply src/config.ts:31]: Check comments bellow.

File: src/utils.ts (hunk 1 above)
Comment (Raiden1411) [reply src/utils.ts:85]: Check comments bellow or above.

**Comment by roninjin10** [comment]
Nice!

**Pull request merged**